### PR TITLE
Add Windows GPU support with Bazel

### PR DESCRIPTION
**Pull request by meteorcloudy**
@mrry @dslomov @damienmg 

The change needed for Bazel is also ready to review:
https://bazel-review.googlesource.com/?polygerrit=0#/c/7351/
Once these changes are checked in, we should be able to build TensorFlow with GPU support on Windows by
`bazel build --confg=win-cuda -c opt --cpu=x64_windows_msvc --cpu=x64_windows tensorflow/tools/pip_package:build_pip_package`

I excluded `depthwise_conv_grad_op` which doesn't seem to be excluded in CMake build.
It caused a linking error like this, and only happens in GPU build:
```
depthwise_conv_grad_op.o : error LNK2019: unresolved external symbol "public: static void __cdecl tensorflow::DepthwiseConv2dBackpropInputGPULaunch<float>::Run(struct Eigen::GpuDevice const &,struct tensorflow::DepthwiseArgs,float const *,float const *,float *)" (?Run@?$DepthwiseConv2dBackpropInputGPULaunch@M@tensorflow@@SAXAEBUGpuDevice@Eigen@@UDepthwiseArgs@2@PEBM2PEAM@Z) referenced in function "public: static void __cdecl tensorflow::LaunchDepthwiseConvBackpropInputOp<struct Eigen::GpuDevice,float>::launch(class tensorflow::OpKernelContext *,struct tensorflow::DepthwiseArgs,float const *,float const *,float *)" (?launch@?$LaunchDepthwiseConvBackpropInputOp@UGpuDevice@Eigen@@M@tensorflow@@SAXPEAVOpKernelContext@2@UDepthwiseArgs@2@PEBM2PEAM@Z)
depthwise_conv_grad_op.o : error LNK2019: unresolved external symbol "public: static void __cdecl tensorflow::DepthwiseConv2dBackpropInputGPULaunch<double>::Run(struct Eigen::GpuDevice const &,struct tensorflow::DepthwiseArgs,double const *,double const *,double *)" (?Run@?$DepthwiseConv2dBackpropInputGPULaunch@N@tensorflow@@SAXAEBUGpuDevice@Eigen@@UDepthwiseArgs@2@PEBN2PEAN@Z) referenced in function "public: static void __cdecl tensorflow::LaunchDepthwiseConvBackpropInputOp<struct Eigen::GpuDevice,double>::launch(class tensorflow::OpKernelContext *,struct tensorflow::DepthwiseArgs,double const *,double const *,double *)" (?launch@?$LaunchDepthwiseConvBackpropInputOp@UGpuDevice@Eigen@@N@tensorflow@@SAXPEAVOpKernelContext@2@UDepthwiseArgs@2@PEBN2PEAN@Z)
depthwise_conv_grad_op.o : error LNK2019: unresolved external symbol "public: static void __cdecl tensorflow::DepthwiseConv2dBackpropFilterGPULaunch<float>::Run(struct Eigen::GpuDevice const &,struct tensorflow::DepthwiseArgs,float const *,float const *,float *)" (?Run@?$DepthwiseConv2dBackpropFilterGPULaunch@M@tensorflow@@SAXAEBUGpuDevice@Eigen@@UDepthwiseArgs@2@PEBM2PEAM@Z) referenced in function "public: static void __cdecl tensorflow::LaunchDepthwiseConvBackpropFilterOp<struct Eigen::GpuDevice,float>::launch(class tensorflow::OpKernelContext *,struct tensorflow::DepthwiseArgs,float const *,float const *,float *)" (?launch@?$LaunchDepthwiseConvBackpropFilterOp@UGpuDevice@Eigen@@M@tensorflow@@SAXPEAVOpKernelContext@2@UDepthwiseArgs@2@PEBM2PEAM@Z)
depthwise_conv_grad_op.o : error LNK2019: unresolved external symbol "public: static void __cdecl tensorflow::DepthwiseConv2dBackpropFilterGPULaunch<double>::Run(struct Eigen::GpuDevice const &,struct tensorflow::DepthwiseArgs,double const *,double const *,double *)" (?Run@?$DepthwiseConv2dBackpropFilterGPULaunch@N@tensorflow@@SAXAEBUGpuDevice@Eigen@@UDepthwiseArgs@2@PEBN2PEAN@Z) referenced in function "public: static void __cdecl tensorflow::LaunchDepthwiseConvBackpropFilterOp<struct Eigen::GpuDevice,double>::launch(class tensorflow::OpKernelContext *,struct tensorflow::DepthwiseArgs,double const *,double const *,double *)" (?launch@?$LaunchDepthwiseConvBackpropFilterOp@UGpuDevice@Eigen@@N@tensorflow@@SAXPEAVOpKernelContext@2@UDepthwiseArgs@2@PEBN2PEAN@Z)
bazel-out/vc_14_0_x64-py3-opt/bin/tensorflow/cc/tutorials_example_trainer.exe : fatal error LNK1120: 4 unresolved externals
Target //tensorflow/cc:tutorials_example_trainer failed to build
INFO: Elapsed time: 2180.664s, Critical Path: 1792.27s
```

The source file [depthwise_conv_grad_op.cc](https://github.com/tensorflow/tensorflow/blob/beb10ceb086fe94a6b1247b45397aafddd47e05d/tensorflow/core/kernels/depthwise_conv_grad_op.cc#L113) looks weird to me, because the syntax highlight starts go wrong from line 113. Might need some help to resolve this.

### CODE REVIEW

**Comment by meteorcloudy** [comment]
There is still an issue that the pip package is trying to load dll libraries with the wrong name on Windows.
Mentioned here: https://github.com/bazelbuild/bazel/issues/2075
I'll fix it after PR #5494 is merged, need to resolve some conflict anyway.

File: configure [hunk 1]
```diff
@@ -184,10 +184,13 @@ while true; do
   fromuser=""
   if [ -z "$CUDA_TOOLKIT_PATH" ]; then
     default_cuda_path=/usr/local/cuda
+    if is_windows; then
+      default_cuda_path="C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v8.0"
```
Comment (mrry) [inline configure]: The CUDA installer seems to have set `%CUDA_PATH%` to this value on my machine. Should we try that location?

File: tensorflow/core/BUILD [hunk 2]
```diff
@@ -908,7 +908,7 @@ cc_library(
         ":lib_internal",
         ":proto_text",
         ":protos_all_cc",
-    ] + if_not_windows([":stream_executor"]),
+    ] + [":stream_executor"],
```
Comment (mrry) [inline tensorflow/core/BUILD]: No need to concatenate the two lists now: can you add `":stream_executor"` to the previous list?

File: tensorflow/core/BUILD [hunk 3]
```diff
@@ -1386,9 +1386,9 @@ tf_cuda_library(
         ":lib_internal",
         ":protos_all_cc",
         "//third_party/eigen3",
-    ] + if_not_windows([
+    ] + [
         ":stream_executor",
-    ]),
+    ],
```
Comment (mrry) [inline tensorflow/core/BUILD]: Can you merge these two lists?

File: tensorflow/python/BUILD [hunk 4]
```diff
@@ -1926,6 +1926,8 @@ tf_cuda_library(
 tf_py_wrap_cc(
     name = "pywrap_tensorflow",
     srcs = ["tensorflow.i"],
+    # set linkstatic = 0 so that only apply whole archive to libraries that need to be force linked.
```
Comment (mrry) [inline tensorflow/python/BUILD]: Can you clarify this comment? (Perhaps mentioning that it (presumably?) is only necessary for Windows.)

**Review by mrry — CHANGES_REQUESTED** [review]

File: tensorflow/core/BUILD (hunk 3 above)
Comment (meteorcloudy) [inline tensorflow/core/BUILD]: Done.

File: configure (hunk 1 above)
Comment (meteorcloudy) [inline configure]: Yes, agree!

**Review by mrry — APPROVED** [review]
LGTM, this is great!

**Comment by meteorcloudy** [comment]
Jenkins, test this please.

**Comment by meteorcloudy** [comment]
Jenkins, test this please.

**Pull request merged**